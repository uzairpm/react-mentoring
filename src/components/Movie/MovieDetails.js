import React, {Component} from 'react';

import Constants from '../Common/constants';
import Header from '../Header/Header';
import Footer from '../Footer/Footer';
import SimilarMovies from './SimilarMovies';
import MovieBanner from './MovieBanner';
import './moviedetails.css';

export default class MovieDetails extends Component {
    constructor(props) {
        super(props);
        this.state = {
            movie: {},
            selectedGenre: '',
            similarMovies: []
        };
    }
    componentDidMount() {
        let url = `${Constants.baseURL}/${this.props.match.params.id}`;
        this.fetchMovieDetails(url).then(response => {
            this.populateState(response);
            this.fetchSimilarMovies(response.genres[0]);
        });
    }
    componentWillReceiveProps(nextProps) {
        let url = `${Constants.baseURL}/${nextProps.match.params.id}`;
        this.fetchMovieDetails(url).then(response => {
            this.populateState(response);
            this.fetchSimilarMovies(response.genres[0]);
        });
    }
    fetchMovieDetails(url) {
        return fetch(url).then(res => res.json());
    }
    populateState(response) {
        this.setState({
            movie: response
        });
    }
    fetchSimilarMovies(genre) {
        let url = `${Constants.baseURL}?search=${genre}&searchBy=genres&sortBy=release_date&sortOrder=desc`;
        fetch(url)
            .then(res => res.json())
            .then(response => {
                this.setState({
                    selectedGenre: genre,
                    similarMovies: response.data
                });
            });
    }
    render() {
        return (
            <div>
                <Header showSearchButton={true} />
                <MovieBanner movie={this.state.movie} />
                <SimilarMovies genre={this.state.selectedGenre} movies={this.state.similarMovies} />
                <Footer />
            </div>
        );
    }
}
