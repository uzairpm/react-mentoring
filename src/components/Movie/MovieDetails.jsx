import React, {Component} from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import Constants from '../Common/constants';
import Header from '../Header/Header';
import Footer from '../Footer/Footer';
import MovieBannerContainer from './containers/MovieBannerContainer';
import SimilarMoviesContainer from './containers/SimilarMoviesContainer';
import * as appActions from '../../actions';

require('es6-promise').polyfill();
require('isomorphic-fetch');

export class MovieDetails extends Component {
    constructor(props) {
        super(props);
    }
    componentDidMount() {
        let url = `${Constants.baseURL}/${this.props.match.params.id}`;
        this.props.fetchMovieDetails(url);
    }
    render() {
        return (
            <div>
                <Header showSearchButton />
                <MovieBannerContainer />
                <SimilarMoviesContainer />
                <Footer />
            </div>
        );
    }
}

const mapDispatchToProps = {
    fetchMovieDetails: appActions.fetchMovieDetails
}

export default withRouter(connect(null, mapDispatchToProps)(MovieDetails));
