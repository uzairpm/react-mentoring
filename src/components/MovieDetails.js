import React, {Component} from 'react';

import Header from './Header/Header';
import Footer from './Footer/Footer';
import SimilarMovies from './SimilarMovies';
import MovieBanner from './MovieBanner';
import './moviedetails.css';

export default class MovieDetails extends Component {
    constructor(props) {
        super(props);
        this.state = {
            showSearchButtonOnHeader: true,
            movie: {},
            selectedGenre: '',
            similarMovies: []
        };
    }
    componentDidMount() {
        let url = 'http://react-cdp-api.herokuapp.com/movies/'+this.props.match.params.id;
        fetch(url)
            .then(res => res.json())
            .then(response => {
                this.setState({
                    movie: response
                });
                this.fetchSimilarMovies(response.genres[0]);
            });
    }
    componentWillReceiveProps(nextProps) {
        let url = 'http://react-cdp-api.herokuapp.com/movies/'+nextProps.match.params.id;
        fetch(url)
            .then(res => res.json())
            .then(response => {
                this.setState({
                    movie: response
                });
                this.fetchSimilarMovies(response.genres[0]);
            });
    }
    fetchSimilarMovies(genre) {
        let url = 'http://react-cdp-api.herokuapp.com/movies?search='+genre+'&searchBy=genres&sortBy=release_date&sortOrder=desc';
        fetch(url)
            .then(res => res.json())
            .then(response => {
                this.setState({
                    selectedGenre: genre,
                    similarMovies: response.data
                });
            });
    }
    render() {
        return (
            <div>
                <Header showSearchButton={this.state.showSearchButtonOnHeader} />
                <MovieBanner movie={this.state.movie} />
                <SimilarMovies genre={this.state.selectedGenre} movies={this.state.similarMovies} />
                <Footer />
            </div>
        );
    }
}
