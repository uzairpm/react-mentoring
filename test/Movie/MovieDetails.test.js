import { shallow } from 'enzyme';
import React from 'react';
const fetchMock = require('fetch-mock');

import MovieDetails from '../../src/components/Movie/MovieDetails';

beforeEach(function() {
    // global.fetch = fetchMock;
});

describe('MovieDetails Component', () => {
    const match = { params: { id: 603 } };
    const element = shallow(<MovieDetails match={match}/>, {disableLifecycleMethods: false});
    test('Snapshot test with default props', () => {
        expect(element).toMatchSnapshot();
    });
    test('Should update state when populateState function is invoked with a movie object', () => {
        var movie = {
            poster_path: '',
            title: 'Matrix',
            vote_average: 7,
            genres: ['Action'],
            release_date : '', 
            runtime: '',
            overview: ''
        };
        element.instance().populateState(movie);
        expect(element.state('movie').title).toEqual('Matrix');
    });
    test('Should update state when populateStateForSimilarMovies function is invoked with movies array', () => {
        var movies = [{genre: ['Action'], id: 100}, {genre: ['Drama'], id: 101}];
        element.instance().populateStateForSimilarMovies('Action', movies);
        expect(element.state('selectedGenre')).toEqual('Action');
    });
    test('Should receive data when request is made using the getRequest function', () => {
        fetchMock.get('*', {hello: 'world'});
        element.instance().getRequest('http://localhost/api')
            .then(function(data) {
                expect(data).toEqual({hello: 'world'});
            });
    });
    test('Should update state when fetchSimilarMovies is invoked with a genre', () => {
        element.instance().fetchSimilarMovies('Action');
        expect(element.state('selectedGenre')).toEqual('Action');
    });
});

afterAll(function() {
    fetchMock.restore();
});
