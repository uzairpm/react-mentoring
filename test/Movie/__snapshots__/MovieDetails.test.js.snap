// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Component Snapshot test with default props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieDetails
    match={
      Object {
        "params": Object {
          "id": 603,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header
          showSearchButton={true}
        />,
        <MovieBanner
          movie={Object {}}
          vote_average={0}
        />,
        <SimilarMovies
          genre=""
          movies={Array []}
        />,
        <Footer />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "showSearchButton": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "movie": Object {},
          "vote_average": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "genre": "",
          "movies": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header
            showSearchButton={true}
          />,
          <MovieBanner
            movie={Object {}}
            vote_average={0}
          />,
          <SimilarMovies
            genre=""
            movies={Array []}
          />,
          <Footer />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "showSearchButton": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "movie": Object {},
            "vote_average": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "genre": "",
            "movies": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": false,
  },
}
`;
